src/util/http_utils.js

```diff
@@ -287,19 +295,66 @@ function send_reply(req, res, reply, options) {
     res.end();
 }
 
+/**
+ * Check if a hostname should be proxied or not
+ */
+function should_proxy(hostname) {
+    for (const addr of no_proxy_list) {
+        // Cover the case where the address is an IP
+        if ((ip.isV4Format(addr) || ip.isV6Format(addr)) && ip.isEqual(hostname, addr)) {
+            return false;
+        }
+        // Cover the case where the address is a CIDR subnet
+        try {
+            if (ip.cidrSubnet(addr).contains(hostname)) {
+                return false;
+            }
+        } catch {
+            // noop, not a cidr subnet
+        }
+        // Cover the case where the address is domain name
+        if (addr.startsWith('.') && hostname.endsWith(addr)) {
+            return false;
+        }
+        // Cover the case where the address is an explicit FQDN
+        if (hostname === addr) {
+            return false;
+        }
+    }
+    return true;
+}
 
 /**
- * Get http / https agent according to protocol
+ * Get http / https agent according to protocol and proxy rules
  */
-function get_unsecured_http_agent(endpoint) {
-    const { protocol } = url.parse(endpoint);
-    // only return unsecured agent for non-aws endpoints
-    if (!cloud_utils.is_aws_endpoint(endpoint)) {
-        return protocol === "https:" ?
-            new https.Agent({
-                rejectUnauthorized: false,
-            }) :
-            new http.Agent();
+function get_http_agent(endpoint, options = {}) {
+    const { HTTP_PROXY, HTTPS_PROXY } = process.env;
+    const { protocol, hostname } = url.parse(endpoint);
+
+    if (protocol === "https:" || protocol === "wss:") {
+        // Do not allow unsecured agents when connecting to aws endpoints
+        if (cloud_utils.is_aws_endpoint(endpoint)) {
+            options.rejectUnauthorized = true;
+        }
+
+        if (HTTPS_PROXY && should_proxy(hostname)) {
+            return new HttpsProxyAgent({
+                ...options,
+                ...url.parse(HTTPS_PROXY),
+            });
+
+        } else {
+            return new https.Agent(options);
+        }
+
+    } else if (HTTP_PROXY && should_proxy(hostname)) {
+        return new HttpProxyAgent({
+            ...options,
+            ...url.parse(HTTP_PROXY)
+        });
+
+    } else {
+        return new http.Agent(options);
```

Comment:
Not sure if an agent is so expensive but I can map to global agents. The issue here will be that the agent options might change for each request so it is not so simple.

Also, the old code also created a new agent for each call, I believe for the same reason.

Reply:
Another Idea is to cache the agent on the call site. in cases where the call site can use the same agent every time.

Reply:
which agent options are we passing per request??

Reply:
The obvious one which is widely used is `rejectUnautorized`, for proxied agents we also set the proxy info.
But as a util I think it might be best to allow agent creation with any custom agent config and still make sure that on proxied evns we get a proxied agent. 

I can suggest the following: 
Return a cached agent in the following cases:
1. non-proxied HTTP agent without custom options
2. proxied HTTP agent without custom options
3. non-proxied non-secured HTTPS agent without custom options
4. proxied non-secured HTTPS agent without custom options
5. non-proxied secured HTTPS agent without custom options
6. proxied secured HTTPS agent without custom options

In any other case create a specific agent and return it.

@guymguym what do you think?

Reply:
Found another gap, we are setting `keepAlive` option  on the global agent, but now we do not use the global agent anymore

Reply:
This is nodejs http.Agent - https://github.com/nodejs/node/blob/master/lib/_http_agent.js
I really don't think this is meant to be created per request, and I don't see why it should.

From my point of view there should be 4 global static agents 1. proxy&secure 2. proxy&unsecure 3. no-proxy&secure 4. no-proxy&unsecure.

And I would like to have two separate exported functions - 1. get_secure_http_agent 2. get_unsecure_http_agent which will return the proper one of these 4.

Reply:
Please refer to my list above, you are missing the other 2 agents, the HTTP one and the PROXIED HTTP one. 
And on top of that, every call site might need it's one instance of an agent when the parameters to the agent differ.
For example `keepAlive=true` vs `keepAlive=false`.

This is why I think caching the agent at the call site is better. you will get a single agent for every invocation from a single call site or even multiple call sites in the same file

Reply:
ok so 6, but even if there are 100 the fact theey are static is important. 
Why do we needs control over keepAlive per request??
I don't think we want to control the agent properties per request, just to pick the agent and make our request with it.

Reply:
I now keep the 6 as static agents and allow to set global settings for all HTTP and HTTPS agents at once using new funciton in http_utils: `update_http_agents` and `update_https_agents`